### PR TITLE
Generate fully qualified table names in the grouping panel according to the SQL dialect

### LINKED ISSUE #35223: [Informix] Grouping panel doesn't work
### Description

Grouping panel doesn't work for informix.
image

### DBeaver Version

24.1.5

### Operating System

win11

### Database and driver

Informix Dynamic Server 14.10.FC9W1DE
Informix JDBC Driver for Informix Dynamic Server 4.50.JC10W1

### Steps to reproduce

_No response_

### Additional context

```
eclipse.buildId=unknown
java.version=17.0.6
java.vendor=Eclipse Adoptium
BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=en
Framework arguments:  -eclipse.keyring C:\Users\Dziyana\AppData\Roaming\DBeaverData\secure\secure_storage
Command-line arguments:  -os win32 -ws win32 -arch x86_64

org.jkiss.dbeaver.model
Error
Thu Aug 15 16:06:52 CEST 2024
A syntax error has occurred.

java.sql.SQLSyntaxErrorException: A syntax error has occurred.
	at com.informix.util.IfxErrMsg.buildExceptionWithMessage(IfxErrMsg.java:416)
	at com.informix.util.IfxErrMsg.buildIsamException(IfxErrMsg.java:403)
	at com.informix.jdbc.IfxSqli.addException(IfxSqli.java:3023)
	at com.informix.jdbc.IfxSqli.receiveError(IfxSqli.java:3274)
	at com.informix.jdbc.IfxSqli.dispatchMsg(IfxSqli.java:2269)
	at com.informix.jdbc.IfxSqli.receiveMessage(IfxSqli.java:2194)
	at com.informix.jdbc.IfxSqli.executeExecute(IfxSqli.java:2104)
	at com.informix.jdbc.IfxResultSet.executeExecute(IfxResultSet.java:302)
	at com.informix.jdbc.IfxStatement.executeImpl(IfxStatement.java:1017)
	at com.informix.jdbc.IfxStatement.executeImpl(IfxStatement.java:997)
	at com.informix.jdbc.IfxStatement.executeBody(IfxStatement.java:745)
	at com.informix.jdbc.IfxStatement.execute(IfxStatement.java:687)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:330)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:131)
	at org.jkiss.dbeaver.ui.controls.resultset.panel.grouping.GroupingDataContainer.readData(GroupingDataContainer.java:129)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:128)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:192)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:126)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$ResultSetDataPumpJob.run(ResultSetViewer.java:5160)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:117)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)


```

### PR DESCRIPTION
**Pull request by ivgag**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java [hunk 1]
```diff
@@ -205,4 +215,57 @@ public String[] getFuncAliases() {
...
+        }
+
+        @Override
+        public void setUnPivot(UnPivot unPivot) {
+            table.setUnPivot(unPivot);
+        }
+
+        @Override
+        public String toString() {
+            String tableName = table.getDatabase()
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java]: Both database and schema are optional.
Also maybe it is better to extend net.sf.jsqlparser.schema.Table than implement interface (besides other issues interface changes in every major version we'll have to reimplement it)

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java [hunk 2]
```diff
@@ -205,4 +211,27 @@ public String[] getFuncAliases() {
...
+            this.sqlDialect = sqlDialect;
+        }
+
+        @Override
+        public String toString() {
+            String databaseName = getDatabase().getDatabaseName() != null ? getDatabase().getDatabaseName() + sqlDialect.getCatalogSeparator() : "";
+            String schemaName = getSchemaName() != null ? getSchemaName() + sqlDialect.getStructSeparator() : "";
+            String fullyQualifiedTableName = databaseName + schemaName + getName();
+            String alias = getAlias() != null ? getAlias().toString() : "";
+            return fullyQualifiedTableName + alias;
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java]: Looks weird. name + alias without space between?

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java [hunk 3]
```diff
@@ -205,4 +211,27 @@ public String[] getFuncAliases() {
...
+        private final SQLDialect sqlDialect;
+
+        public FormattedTable(Table table, SQLDialect sqlDialect) {
+            super(table.getDatabase(), table.getSchemaName(), table.getName());
+            this.sqlDialect = sqlDialect;
+        }
+
+        @Override
+        public String toString() {
+            String databaseName = getDatabase().getDatabaseName() != null ? getDatabase().getDatabaseName() + sqlDialect.getCatalogSeparator() : "";
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java]: Potentially it is possible that database name will be empty (""). Use CommonUtils.isEmpty

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java (hunk 2 above)
Comment (serge-rider) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java]: Perhaps it is better to use StringBuilder here.

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java (hunk 2 above)
Comment (ivgag) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java]: `net.sf.jsqlparser.expression.Alias#toString` already adds a space before the alias.

Comment (ivgag) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java]: Do we really consider this a bottleneck that justifies sacrificing readability for performance?

Comment (serge-rider) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java]: No it is just weird looking code. Relying on undocumented feature of 3rd party library doesn't sound very good.

Performance here doesn't really matter, it is about readability.

Comment (ivgag) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/SQLGroupingQueryGenerator.java]: Yeah, agree.

I tried to do that using SB, but I found the result even less readable

```
@Override
public String toString() {
    StringBuilder fullyQualifiedTableName = new StringBuilder();
    
    if (getDatabase().getDatabaseName() != null) {
        fullyQualifiedTableName.append(getDatabase().getDatabaseName())
                               .append(sqlDialect.getCatalogSeparator());
    }
    if (getSchemaName() != null) {
        fullyQualifiedTableName.append(getSchemaName())
                               .append(sqlDialect.getStructSeparator());
    }
    
    fullyQualifiedTableName.append(getName());
    
    if (getAlias() != null) {
        fullyQualifiedTableName.append(getAlias().toString());
    }
    
    return fullyQualifiedTableName.toString();
}

```

**Comment by uslss** [comment]
verified

**Pull request merged**